cmd/topicctl/subcmd/rebalance.go

```diff
@@ -0,0 +1,323 @@
+
+	shared sharedOptions
+}
+
+var rebalanceConfig rebalanceCmdConfig
+
+func init() {
+	rebalanceCmd.Flags().IntSliceVar(
+		&rebalanceConfig.brokersToRemove,
+		"to-remove",
+		[]int{},
+		"Brokers to remove; only applies if rebalance is also set",
+	)
+	rebalanceCmd.Flags().IntVar(
+		&rebalanceConfig.brokerThrottleMBsOverride,
+		"broker-throttle-mb",
+		0,
+		"Broker throttle override (MB/sec)",
+	)
+	rebalanceCmd.Flags().BoolVar(
+		&rebalanceConfig.dryRun,
+		"dry-run",
+		false,
+		"Do a dry-run",
+	)
+	rebalanceCmd.Flags().IntVar(
+		&rebalanceConfig.partitionBatchSizeOverride,
+		"partition-batch-size",
+		0,
+		"Partition batch size override",
+	)
+	rebalanceCmd.Flags().StringVar(
+		&rebalanceConfig.pathPrefix,
+		"path-prefix",
+		os.Getenv("TOPICCTL_APPLY_PATH_PREFIX"),
+		"Prefix for topic config paths",
+	)
+	rebalanceCmd.Flags().DurationVar(
+		&rebalanceConfig.sleepLoopDuration,
+		"sleep-loop-duration",
+		10*time.Second,
+		"Amount of time to wait between partition checks",
+	)
+	rebalanceCmd.Flags().DurationVar(
+		&rebalanceConfig.showProgressInterval,
+		"show-progress-interval",
+		0*time.Second,
+		"show progress during rebalance at intervals",
+	)
+
+	addSharedConfigOnlyFlags(rebalanceCmd, &rebalanceConfig.shared)
+	RootCmd.AddCommand(rebalanceCmd)
+}
+
+func rebalancePreRun(cmd *cobra.Command, args []string) error {
+	if rebalanceConfig.shared.clusterConfig == "" || rebalanceConfig.pathPrefix == "" {
+		return fmt.Errorf("Must set args --cluster-config & --path-prefix (or) env variables TOPICCTL_CLUSTER_CONFIG & TOPICCTL_APPLY_PATH_PREFIX")
+	}
+
+	return nil
+}
+
+func rebalanceRun(cmd *cobra.Command, args []string) error {
+	ctx := context.Background()
+	rebalanceCtxMap, err := getRebalanceCtxMap(&rebalanceConfig)
+	if err != nil {
+		return err
+	}
+	ctx = context.WithValue(ctx, "progress", rebalanceCtxMap)
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, os.Interrupt, syscall.SIGTERM)
+	go func() {
+		// for an interrupt, cancel context and exit program to end all topic rebalances
+		<-sigChan
+		cancel()
+		os.Exit(1)
+	}()
+
+	clusterConfigPath := rebalanceConfig.shared.clusterConfig
+	topicConfigDir := rebalanceConfig.pathPrefix
+	clusterConfig, err := config.LoadClusterFile(clusterConfigPath, rebalanceConfig.shared.expandEnv)
+	if err != nil {
+		return err
+	}
+
+	adminClient, err := clusterConfig.NewAdminClient(ctx,
+		nil,
+		rebalanceConfig.dryRun,
+		rebalanceConfig.shared.saslUsername,
+		rebalanceConfig.shared.saslPassword,
+	)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer adminClient.Close()
+
+	// get all topic configs from path-prefix i.e topics folder
+	// this folder should only have topic configs
+	// recursive search is not performed in the path-prefix
+	log.Infof("Getting all topic configs from path prefix (from root) %v", topicConfigDir)
+	matches, err := filepath.Glob(topicConfigDir + "/**")
```

Comment:
> this folder should only have topic configs 

Is this a reasonable assumption? Does it make sense to filter on *.yaml or are you looking to bail on errors loading non-yaml files below?

Reply:
If we perform `topicctl apply $TOPIC_NAME` with --cluster-config & --path-prefix, it expects a
1. path with topic name as "topic: blah" where topic name do not have ".yaml" extension
2. cluster yaml file with cluster file extension as yaml i.e cluster.yaml

Example:
file name of topic: test-2 is test-2
file name of topic: test-4 it test-4.yaml

<img width="1029" alt="Screenshot 2023-06-07 at 21 38 07" src="https://github.com/segmentio/topicctl/assets/105226401/dad14232-858f-4049-bca3-e06b5ed188df">

## go run cmd/topicctl/main.go apply test-2 --skip-confirm --partition-batch-size=10 --cluster-config /path/to/cluster/cluster.yaml --path-prefix /path/to/cluster/topics/
<img width="1640" alt="Screenshot 2023-06-07 at 21 42 03" src="https://github.com/segmentio/topicctl/assets/105226401/4b2eec05-eb33-4f00-96b5-a8578ef8609d">

## go run cmd/topicctl/main.go apply test-4 --skip-confirm --partition-batch-size=10 --cluster-config /path/to/cluster/cluster.yaml --path-prefix /path/to/cluster/topics/
<img width="1640" alt="Screenshot 2023-06-07 at 21 42 45" src="https://github.com/segmentio/topicctl/assets/105226401/e32d60ed-d736-49fa-94a0-e69efa40920a">

If you notice test-4 topic is not found in `topicctl apply`

Now I renamed test-4.yaml to test-4 and it works!!
<img width="1035" alt="Screenshot 2023-06-07 at 21 44 11" src="https://github.com/segmentio/topicctl/assets/105226401/9d1e3116-84cf-48c4-b12e-4693fa3b6946">

## go run cmd/topicctl/main.go apply test-4 --skip-confirm --partition-batch-size=10 --cluster-config /path/to/cluster/cluster.yaml --path-prefix /path/to/cluster/topics/
<img width="1637" alt="Screenshot 2023-06-07 at 21 45 02" src="https://github.com/segmentio/topicctl/assets/105226401/416b3143-4ce1-427e-a2da-f69247f46b13">

To ensure we adhere to `topicctl apply` conventions, I had to follow this way.

I hope this is clear? or am i missing something?

Note: `topictl rebalance` leverages topicctl apply to rebalance a topic

Reply:
Update to above.. this has been fixed

Current changes will
- recursively go through all files for a given --path-prefix
- If a cluster.yaml is found, it is ignored
- for each topic, if topic is not in sync with cluster.yaml, it will log error and continue

Can this be resolved?